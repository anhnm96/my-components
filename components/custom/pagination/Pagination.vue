<script lang="ts" setup>
// import { ChevronLeftIcon, ChevronRightIcon } from '@heroicons/vue/outline'

const props = defineProps({
  currentPage: {
    type: Number,
    default: 1,
  },
  total: {
    type: Number,
    required: true,
  },
  perPage: {
    type: Number,
    default: 20,
  },
  rangeBefore: {
    type: Number,
    default: 1,
  },
  rangeAfter: {
    type: Number,
    default: 1,
  },
  ariaNextLabel: {
    type: String,
    default: 'next page',
  },
  ariaPreviousLabel: {
    type: String,
    default: 'previous page',
  },
  ariaPageLabel: {
    type: String,
    default: 'page',
  },
  ariaCurrentLabel: {
    type: String,
    default: 'current',
  },
})

const emit = defineEmits(['update:currentPage', 'change'])
const pageCount = computed(() => Math.ceil(props.total / props.perPage))
const hasPrev = computed(() => {
  return props.currentPage > 1
})
const hasFirst = computed(() => {
  return props.currentPage >= 2 + props.rangeBefore
})
const hasFirstEllipsis = computed(() => {
  return props.currentPage >= props.rangeBefore + 4
})
const hasLast = computed(() => {
  return props.currentPage <= pageCount.value - (1 + props.rangeAfter)
})
const hasLastEllipsis = computed(() => {
  return props.currentPage < pageCount.value - (2 + props.rangeAfter)
})
const hasNext = computed(() => {
  return props.currentPage < pageCount.value
})
const pagesInRange = computed(() => {
  let left = Math.max(1, props.currentPage - props.rangeBefore)
  if (left - 1 === 2) {
    // Do not show the ellipsis if there is only one to hide
    left--
  }
  let right = Math.min(props.currentPage + props.rangeAfter, pageCount.value)
  if (pageCount.value - right === 2) {
    // Do not show the ellipsis if there is only one to hide
    right++
  }
  const pages = []
  for (let i = left; i <= right; i++) pages.push(getPage(i))

  return pages
})

function changePage(event: MouseEvent, pageNumber: number) {
  // Do nothing if clicking the current active page
  if (pageNumber === props.currentPage) return

  emit('update:currentPage', pageNumber)
  emit('change', pageNumber)

  // Set focus on element to keep tab order
  if (event.target) nextTick(() => (event.target as HTMLElement).focus())
}
function getPage(number: number) {
  return {
    number,
    isCurrent: props.currentPage === number,
    onClick: (event: MouseEvent) => changePage(event, number),
    ariaLabel:
      props.currentPage === number
        ? `${props.ariaCurrentLabel}, ${props.ariaPageLabel} ${number}`
        : `${props.ariaPageLabel} ${number}`,
  }
}
</script>

<template>
  <nav
    class="relative z-0 inline-flex space-x-3 rounded-md shadow-sm"
    aria-label="pagination"
  >
    <button
      :aria-label="`${ariaPreviousLabel}, ${ariaPageLabel} ${currentPage - 1}`"
      class="btn-page"
      :disabled="!hasPrev"
      data-test="hasPrev"
      @click="changePage($event, currentPage - 1)"
    >
      <!-- <ChevronLeftIcon class="w-4 h-4" aria-hidden="true" /> -->
      &lt;
    </button>
    <button
      v-if="hasFirst"
      class="btn-page"
      :aria-label="`${props.ariaPageLabel} 1`"
      data-test="hasFirst"
      @click="changePage($event, 1)"
    >
      1
    </button>
    <button v-if="hasFirstEllipsis" class="btn-page">&hellip;</button>
    <!-- Pages -->
    <button
      v-for="page in pagesInRange"
      :key="page.number"
      :aria-label="page.ariaLabel"
      :aria-current="page.isCurrent"
      class="btn-page"
      :class="[page.isCurrent && 'btn-page--active']"
      :data-test="`page-${page.number}`"
      @click="page.onClick"
    >
      {{ page.number }}
    </button>
    <button v-if="hasLastEllipsis" class="btn-page">&hellip;</button>
    <button
      v-if="hasLast"
      class="btn-page"
      :aria-label="`${props.ariaPageLabel} ${pageCount}`"
      data-test="hasLast"
      @click="changePage($event, pageCount)"
    >
      {{ pageCount }}
    </button>
    <button
      :aria-label="`${ariaNextLabel}, ${ariaPageLabel} ${currentPage + 1}`"
      :disabled="!hasNext"
      class="btn-page"
      data-test="hasNext"
      @click="changePage($event, currentPage + 1)"
    >
      <!-- <ChevronRightIcon class="w-4 h-4" aria-hidden="true" /> -->
      &gt;
    </button>
  </nav>
</template>

<style scoped>
.btn-page {
  @apply inline-flex h-9 w-9 items-center justify-center text-gray-800;
}

.btn-page--active {
  @apply rounded-full bg-gray-100 text-gray-600;
}
</style>
