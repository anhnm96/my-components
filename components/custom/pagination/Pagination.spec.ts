import type { DOMWrapper, VueWrapper } from '@vue/test-utils'
import { shallowMount } from '@vue/test-utils'
import Pagination from './Pagination.vue'

let wrapper: VueWrapper

function testCurrentPageButton(page: Omit<DOMWrapper<Element>, 'exists'>) {
  expect(page.classes()).toContain('btn-page--active')
  expect(page.attributes('aria-current')).toBe('true')
}

describe('Pagination', () => {
  beforeEach(() => {
    wrapper = shallowMount(Pagination, {
      props: {
        total: 200,
        rangeBefore: 2,
        rangeAfter: 2,
        'onUpdate:currentPage': (value: number) =>
          wrapper.setProps({ currentPage: value }),
      },
    }) as VueWrapper
  })

  it('render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should go to page', async () => {
    const page = wrapper.get('[data-test="page-3"]')
    await page.trigger('click')

    expect(wrapper.emitted('change')).toEqual([[3]])

    testCurrentPageButton(page)
  })

  it('should start from currentPage value', async () => {
    await wrapper.setProps({ currentPage: 5 })

    const page = wrapper.get('[data-test="page-5"]')
    testCurrentPageButton(page)
  })

  it('should go next', async () => {
    await wrapper.setProps({ currentPage: 5 })
    const btnNext = wrapper.get('[data-test="hasNext"]')
    await btnNext.trigger('click')
    await btnNext.trigger('click')
    const changeEvent = wrapper.emitted('change')
    expect(changeEvent).toHaveLength(2)
    expect(changeEvent?.[0]).toEqual([6])
    expect(changeEvent?.[1]).toEqual([7])

    const page = wrapper.get('[data-test="page-7"]')
    testCurrentPageButton(page)
  })

  it('should go previous', async () => {
    await wrapper.setProps({ currentPage: 5 })
    const btnPrevious = wrapper.get('[data-test="hasPrev"]')
    await btnPrevious.trigger('click')
    await btnPrevious.trigger('click')
    const changeEvent = wrapper.emitted('change')
    expect(changeEvent).toHaveLength(2)
    expect(changeEvent?.[0]).toEqual([4])
    expect(changeEvent?.[1]).toEqual([3])

    const page = wrapper.get('[data-test="page-3"]')
    testCurrentPageButton(page)
  })

  it('should not go prev when at first page and could go to last page', async () => {
    const btnPrev = wrapper.get('[data-test="hasPrev"]')
    expect(btnPrev.attributes('disabled')).toBe('')
    // go to last page
    await wrapper.get('[data-test="hasLast"]').trigger('click')
    const page = wrapper.get('[data-test="page-10"]')
    testCurrentPageButton(page)
  })

  it('should not go next when at last page and could go to first page', async () => {
    await wrapper.setProps({ currentPage: 10 })
    const btnNext = wrapper.get('[data-test="hasNext"]')
    expect(btnNext.attributes('disabled')).toBe('')
    // go to first page
    await wrapper.get('[data-test="hasFirst"]').trigger('click')
    const page = wrapper.get('[data-test="page-1"]')
    testCurrentPageButton(page)
  })
})
