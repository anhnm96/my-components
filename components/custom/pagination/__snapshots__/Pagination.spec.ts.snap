// Vitest Snapshot v1

exports[`Pagination > render correctly 1`] = `
"<nav class=\\"relative z-0 inline-flex space-x-3 rounded-md shadow-sm\\" aria-label=\\"pagination\\" data-v-88728be1=\\"\\"><button aria-label=\\"previous page, page 0\\" class=\\"btn-page\\" disabled=\\"\\" data-test=\\"hasPrev\\" data-v-88728be1=\\"\\">
    <!-- <ChevronLeftIcon class=\\"w-4 h-4\\" aria-hidden=\\"true\\" /> -->
    < </button>
      <!--v-if-->
      <!--v-if-->
      <!-- Pages --><button aria-label=\\"current, page 1\\" aria-current=\\"true\\" class=\\"btn-page btn-page--active\\" data-test=\\"page-1\\" data-v-88728be1=\\"\\">1</button><button aria-label=\\"page 2\\" aria-current=\\"false\\" class=\\"btn-page\\" data-test=\\"page-2\\" data-v-88728be1=\\"\\">2</button><button aria-label=\\"page 3\\" aria-current=\\"false\\" class=\\"btn-page\\" data-test=\\"page-3\\" data-v-88728be1=\\"\\">3</button><button class=\\"btn-page\\" data-v-88728be1=\\"\\">â€¦</button><button class=\\"btn-page\\" aria-label=\\"page 10\\" data-test=\\"hasLast\\" data-v-88728be1=\\"\\">10</button><button aria-label=\\"next page, page 2\\" class=\\"btn-page\\" data-test=\\"hasNext\\" data-v-88728be1=\\"\\">
        <!-- <ChevronRightIcon class=\\"w-4 h-4\\" aria-hidden=\\"true\\" /> --> >
      </button></nav>"
`;
